lib/backend/k8s/syncer.go

```diff
@@ -0,0 +1,536 @@
+}
+
+func (syn *kubeSyncer) Start() {
+	// Channel for receiving updates from a snapshot.
+	snapshotUpdates := make(chan *[]model.Update)
+
+	// Channel for receiving updates from the watcher.
+	watchUpdates := make(chan *model.Update)
+
+	// Channel used by the watcher to trigger a re-sync.
+	triggerResync := make(chan *resourceVersions, 5)
+
+	// Channel to send the index from which to start the snapshot.
+	initialSnapshotIndex := make(chan *resourceVersions)
+
+	// Status channel
+	statusUpdates := make(chan api.SyncStatus)
+
+	// If we're not in one-shot mode, start the API watcher to
+	// gather updates.
+	if !syn.OneShot {
+		go syn.watchKubeAPI(watchUpdates, triggerResync, initialSnapshotIndex)
+	}
+
+	// Start a background thread to read snapshots from etcd.  It will
+	// read a start-of-day snapshot and then wait to be signalled on the
+	// resyncIndex channel.
+	go syn.readSnapshot(snapshotUpdates, triggerResync, initialSnapshotIndex, statusUpdates)
+
+	// Start a routine to merge updates from the snapshot routine and the
+	// watch routine (if running), and pass information to callbacks.
+	// TODO: We're never actually running both the snapshot thread and the watch thread at the same time
+	// so what's the point?
+	go syn.mergeUpdates(snapshotUpdates, watchUpdates, statusUpdates)
+}
+
+// mergeUpdates ensures that callbacks are all executed from the same thread.
+func (syn *kubeSyncer) mergeUpdates(snapshotUpdates chan *[]model.Update, watchUpdates chan *model.Update, statusUpdates chan api.SyncStatus) {
+	var update *model.Update
+	var updates *[]model.Update
+	currentStatus := api.WaitForDatastore
+	newStatus := api.WaitForDatastore
+	syn.callbacks.OnStatusUpdated(api.WaitForDatastore)
+	for {
+		select {
+		case updates = <-snapshotUpdates:
+			log.Debugf("Snapshot update: %+v", updates)
+			if currentStatus != api.ResyncInProgress {
+				panic("Recieved snapshot update while not resyncing")
+			}
+			syn.callbacks.OnUpdates(*updates)
+		case update = <-watchUpdates:
+			log.Debugf("Watch update: %+v", update)
+			if currentStatus != api.InSync {
+				panic("Recieved watch update while not in sync")
+			}
+			syn.callbacks.OnUpdates([]model.Update{*update})
+		case newStatus = <-statusUpdates:
+			if newStatus != currentStatus {
+				log.Debugf("Status update. %s -> %s", currentStatus, newStatus)
+				syn.callbacks.OnStatusUpdated(newStatus)
+				currentStatus = newStatus
+			}
+		}
+	}
+}
+
+func (syn *kubeSyncer) readSnapshot(updateChan chan *[]model.Update,
+	resyncChan chan *resourceVersions, initialVersionChan chan *resourceVersions, statusUpdates chan api.SyncStatus) {
+
+	log.Info("Starting readSnapshot worker")
+	// Indicate we're starting a resync.
+	statusUpdates <- api.ResyncInProgress
+
+	// Perform an initial snapshot, and send the latest versions to the
+	// watcher routine.
+	initialVersions := resourceVersions{}
+	updateChan <- syn.performSnapshot(&initialVersions)
+
+	// Indicate we're in sync for the first time.
```

Comment:
Thought there might be a race here if the InSync message gets delayed but looks like you avoid that because the channel isn't buffered. IIUC, that means the send is a rendezvous point for the two threads so you know the InSync was received before you send the initialVersions to the watcher.

Probably worth a comment that you're relying on statusUpdates being unbuffered?